### PR TITLE
Fix CCE when using `foreach` with `xml<xml<T>>`

### PR DESCRIPTION
**Pull request by MaryamZi**
## Purpose
$title.

Fixes https://github.com/ballerina-platform/ballerina-lang/issues/35280

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/jballerina-unit-test/src/test/resources/test-src/types/xml/xml_iteration.bal [hunk 1]
```diff
@@ -290,6 +290,19 @@ function xmlTypeParamPIIter() {
     assert(nextPI2.toString(), "{\"value\":`<?target data?>`}");
 }
 
+function testSequenceOfSequenceOfXmlElementIteration() {
+    xml<xml<xml:Element>> elements = xml `<foo/><bar>value</bar><baz>1</baz>`;
```
Comment (gimantha) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/xml/xml_iteration.bal:294]: Can this be a valid test case ?
```ballerina
type xml:Element XML;
xml<xml<XML>> elements = xml `<foo/><bar>value</bar><baz>1</baz>`;
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java [hunk 2]
```diff
@@ -1689,7 +1689,7 @@ public void setForeachTypedBindingPatternType(BLangForeach foreachNode) {
                 BType constraint = getReferredType(((BXMLType) collectionType).constraint);
                 while (constraint.tag == TypeTags.XML) {
                     collectionType = constraint;
-                    constraint = ((BXMLType) collectionType).constraint;
+                    constraint = getReferredType(((BXMLType) collectionType).constraint);
```
Comment (chiranSachintha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java:1692]: Shall we refactor the logic in the `[getReferredType](L1309)` function as follows
```java
public static BType getReferredType(BType type) {
        if (type.tag == TypeTags.TYPEREFDESC) {
            return getReferredType(((BTypeReferenceType) type).referredType);
        }
        return type;
    }
```

**Comment by chiranSachintha** [comment]
For the following kind of scenario still we get the CCE
```ballerina
public function main() {
    xml<xml<xml:Element & readonly>> elements = xml ``;

    foreach var element in elements {
        
    }
}
```

**Comment by MaryamZi** [comment]
> For the following kind of scenarios still we get the CCE
> 
> ```ballerina
> // Example 1
> public function main() {
>     xml<xml<xml:Element & readonly>> elements = xml ``;
> 
>     foreach var element in elements {
>         
>     }
> }
> 
> // Example 2
> public function main() {
>     xml<xml<xml:Element[]>> elements = xml ``;
> 
>     foreach var element in elements {
>         
>     }
> }
> ```

These are separate issues, right? Also reproducible with

```ballerina
public function main() {
    xml<xml:Element & readonly> elements = xml ``;

    foreach var element in elements {
        
    }
}
```

```ballerina
public function main() {
    xml<xml:Element[]> elements = xml ``;

    foreach var element in elements {
        
    }
}
```

I think we can fix them separately. Can you create an issue?

**Comment by chiranSachintha** [comment]
Create an issue(https://github.com/ballerina-platform/ballerina-lang/issues/35473) to track  https://github.com/ballerina-platform/ballerina-lang/pull/35438#issuecomment-1067730544

**Pull request merged**